File: kid-ios/MyWorld/Parent/StoreView.swift
import SwiftUI
import StoreKit

/// Credits & Store — the iOS-compliant purchase surface.
///
/// APPLE POLICY: credits and the subscription are digital goods, so inside the
/// app they are sold ONLY through StoreKit in-app purchase (consumable packs +
/// an auto-renewable subscription). No external purchase links here. After a
/// verified purchase we post the signed transaction to /api/store?action=
/// iap-verify, which grants the credits to the parent's cross-platform wallet
/// (idempotent per transaction id — safe to re-send on relaunch).
///
/// Product ids must exist in App Store Connect:
///   credits50/100/250/500/1000 (consumables);
///   starter.monthly / plus.monthly / pro.monthly (auto-renew, ONE subscription
///   group so Apple handles upgrades/downgrades between tiers natively).
struct StoreView: View {
    @Environment(AuthManager.self) private var auth

    @State private var products: [Product] = []
    @State private var balance: Int?
    @State private var busy: String?          // product id mid-purchase
    @State private var note: String?
    @State private var loadFailed = false
    @State private var couponCode = ""
    @State private var redeeming = false
    @State private var entitlement: APIClient.StoreEntitlement?

    // Starter is retired at launch (parent feedback — relaunching later is a
    // server-side unhide + re-adding the id here + an ASC product).
    private static let productIDs = ["plus.monthly", "pro.monthly",
                                     "credits50", "credits100", "credits250", "credits500", "credits1000"]
    private let api = APIClient()

    private var memberships: [Product] { products.filter { $0.type == .autoRenewable }.sorted { $0.price < $1.price } }
    private var packs: [Product] { products.filter { $0.type != .autoRenewable }.sorted { $0.price < $1.price } }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 18) {
                header

                Text("Every image you make is your family's to keep, stored safely forever, even when you change or regenerate one.")
                    .font(.system(size: 14))
                    .foregroundStyle(.secondary)
                    .padding(14)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .background(Color(hex: "#fff7fb"))
                    .clipShape(RoundedRectangle(cornerRadius: 14))

                Text("1 credit makes one picture tile. A family-member portrait uses 5 (it runs on our best likeness model). Every image includes one free retry.")
                    .font(.system(size: 13))
                    .foregroundStyle(.secondary)

                if products.isEmpty {
                    if loadFailed {
                        VStack(alignment: .leading, spacing: 10) {
                            Text("The store couldn't load products right now. This can happen on a weak connection, or while the App Store catalog is updating.")
                                .font(.system(size: 14)).foregroundStyle(.secondary)
                            Button {
                                Task { await load() }
                            } label: {
                                Text("Try again")
                                    .font(.system(size: 14, weight: .bold))
                                    .padding(.horizontal, 16).padding(.vertical, 9)
                                    .background(Color(hex: "#ff1493"))
                                    .foregroundStyle(.white)
                                    .clipShape(Capsule())
                            }
                            .buttonStyle(.plain)
                        }
                    } else {
                        ProgressView().frame(maxWidth: .infinity)
                    }
                } else {
                    if !memberships.isEmpty {
                        Text("MEMBERSHIPS")
                            .font(.system(size: 12, weight: .heavy))
                            .foregroundStyle(Color(hex: "#ad1457"))
                        Text("Every membership unlocks speech-to-text mode, automatic teaching, reporting, and data saving. They differ in monthly image credits and voice budget.")
                            .font(.system(size: 12)).foregroundStyle(.secondary)
                        ForEach(memberships, id: \.id) { p in
                            membershipRow(p)
                        }
                        // Upgrades/downgrades/cancel live in Apple's own sheet.
                        Link("Manage subscription", destination: URL(string: "https://apps.apple.com/account/subscriptions")!)
                            .font(.system(size: 13, weight: .semibold))
                    }
                    Text("CREDIT PACKS")
                        .font(.system(size: 12, weight: .heavy))
                        .foregroundStyle(Color(hex: "#ad1457"))
                        .padding(.top, 4)
                    if entitlement?.tier == "free" {
                        // Packs top up a MEMBERSHIP — every styled spend
                        // checks membership before the wallet, so selling a
                        // free-tier parent credits they can't use would be
                        // taking money for nothing. Point at memberships.
                        Text("Credit packs top up a membership. Join My World Plus or Pro above first, then packs stack on top of your monthly ⭐.")
                            .font(.system(size: 13))
                            .foregroundStyle(.secondary)
                            .padding(12)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Color(hex: "#fff7fb"), in: RoundedRectangle(cornerRadius: 12))
                    } else {
                        ForEach(packs, id: \.id) { p in
                            productRow(p)
                        }
                    }
                }

                if let n = note {
                    Text(n).font(.system(size: 13, weight: .semibold))
                        .foregroundStyle(Color(hex: "#047857"))
                }

                // Coupon redeem — credits granted by a code (family gift,
                // therapist drop, promo). One use per account, server-enforced.
                HStack(spacing: 10) {
                    TextField("Have a code?", text: $couponCode)
                        .textInputAutocapitalization(.characters)
                        .autocorrectionDisabled()
                        .font(.system(size: 15, weight: .semibold, design: .rounded))
                        .padding(.horizontal, 12).padding(.vertical, 10)
                        .background(Color(.systemBackground))
                        .clipShape(RoundedRectangle(cornerRadius: 12))
                        .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color(hex: "#f3c6dd"), lineWidth: 2))
                    Button {
                        Task { await redeem() }
                    } label: {
                        Text(redeeming ? "…" : "Redeem")
                            .font(.system(size: 14, weight: .bold))
                            .padding(.horizontal, 16).padding(.vertical, 10)
                            .background(Color(hex: "#ad1457"))
                            .foregroundStyle(.white)
                            .clipShape(Capsule())
                    }
                    .buttonStyle(.plain)
                    .disabled(redeeming || couponCode.trimmingCharacters(in: .whitespaces).isEmpty)
                }

                NavigationLink { WordShopView() } label: {
                    HStack {
                        Image(systemName: "cart.fill")
                        Text("Shop words for the board")
                            .font(.system(size: 15, weight: .bold, design: .rounded))
                        Spacer()
                        Image(systemName: "chevron.right").font(.system(size: 13, weight: .bold))
                    }
                    .padding(14)
                    .background(Color(hex: "#fff7fb"))
                    .foregroundStyle(Color(hex: "#ad1457"))
                    .clipShape(RoundedRectangle(cornerRadius: 14))
                    .overlay(RoundedRectangle(cornerRadius: 14).stroke(Color(hex: "#f3c6dd"), lineWidth: 2))
                }
                .buttonStyle(.plain)

                Button("Restore purchases") { Task { await restore() } }
                    .font(.system(size: 13))
                    .padding(.top, 6)

                // Apple Review 3.1.2: the auto-renew terms + working links to
                // the Terms and Privacy Policy must live in the app.
                VStack(alignment: .leading, spacing: 8) {
                    Text("Joining builds your child's personalized starter board right away, 100+ core words plus two family portraits, charged at the lower of the build's credit price or your monthly grant; Plus invests its whole first month in the build, and Pro always finishes enrollment with at least ⭐50 remaining. Cancel anytime: everything you've made stays yours, forever. Memberships: My World Plus is $9.99/month (50 credits monthly) and My World Pro is $19.99/month (150 credits monthly). Payment is charged to your Apple ID at confirmation of purchase. Subscriptions renew automatically unless cancelled at least 24 hours before the end of the current period. Manage, switch tiers, or cancel any time in Settings → Apple ID → Subscriptions. Credits are non-refundable once spent on completed images, and every image you make is yours to keep.")
                        .font(.system(size: 11))
                        .foregroundStyle(.secondary)
                    HStack(spacing: 14) {
                        Link("Terms of Service", destination: URL(string: "\(APIClient.defaultOrigin)/terms")!)
                        Link("Privacy Policy", destination: URL(string: "\(APIClient.defaultOrigin)/privacy")!)
                    }
                    .font(.system(size: 12, weight: .semibold))
                }
                .padding(.top, 10)
            }
            .padding(18)
        }
        .navigationTitle("Credits & Store")
        .task { await load() }
        // Warm the Word Shop's catalog cache while the parent is still on this
        // screen, so tapping "Shop words for the board" opens instantly.
        .task { await ShopCatalog.refresh(childId: auth.childSlug) }
    }

    private var header: some View {
        HStack {
            VStack(alignment: .leading, spacing: 2) {
                Text("Your credits").font(.system(size: 13)).foregroundStyle(.secondary)
                Text(balance.map { "⭐ \($0)" } ?? "⭐ …")
                    .font(.system(size: 30, weight: .heavy, design: .rounded))
                    .foregroundStyle(Color(hex: "#ad1457"))
            }
            Spacer()
            if let e = entitlement {
                VStack(alignment: .trailing, spacing: 2) {
                    Text("Your plan").font(.system(size: 13)).foregroundStyle(.secondary)
                    Text(e.label)
                        .font(.system(size: 15, weight: .heavy, design: .rounded))
                        .foregroundStyle(e.tier == "free" ? Color.secondary : Color(hex: "#047857"))
                }
            }
        }
    }

    private func membershipBlurb(_ id: String) -> String {
        switch id {
        case "starter.monthly": return "10 image credits/month · speech-to-text · auto-teach · reporting"
        case "plus.monthly":    return "joins with your whole starter board personalized up front (a ⭐120+ value) · ⭐50/month · speech-to-text · auto-teach · reporting"
        case "pro.monthly":     return "same enrollment build, and ⭐50 still yours after sign-up · ⭐150/month · biggest voice budget · new features first"
        default:                return "Membership"
        }
    }

    @ViewBuilder
    private func membershipRow(_ p: Product) -> some View {
        let isCurrent = entitlement?.tier == p.id
        HStack(spacing: 14) {
            VStack(alignment: .leading, spacing: 2) {
                HStack(spacing: 6) {
                    Text(p.displayName.isEmpty ? p.id : p.displayName)
                        .font(.system(size: 16, weight: .bold, design: .rounded))
                    if isCurrent {
                        Text("CURRENT")
                            .font(.system(size: 9, weight: .heavy))
                            .padding(.horizontal, 6).padding(.vertical, 2)
                            .background(Capsule().fill(Color(hex: "#ecfdf5")))
                            .foregroundStyle(Color(hex: "#047857"))
                    }
                }
                Text(membershipBlurb(p.id))
                    .font(.system(size: 12)).foregroundStyle(.secondary)
            }
            Spacer()
            Button {
                Task { await buy(p) }
            } label: {
                if busy == p.id { ProgressView() }
                else {
                    Text(isCurrent ? "Yours" : "\(p.displayPrice)/mo")
                        .font(.system(size: 14, weight: .bold))
                        .padding(.horizontal, 16).padding(.vertical, 9)
                        .background(isCurrent ? Color(hex: "#f3c6dd") : Color(hex: "#ff1493"))
                        .foregroundStyle(.white)
                        .clipShape(Capsule())
                }
            }
            .buttonStyle(.plain)
            .disabled(busy != nil || isCurrent)
        }
        .padding(14)
        .background(Color(.systemBackground))
        .clipShape(RoundedRectangle(cornerRadius: 14))
        .overlay(RoundedRectangle(cornerRadius: 14)
            .stroke(isCurrent ? Color(hex: "#047857") : Color(hex: "#f3c6dd"), lineWidth: 2))
    }

    @ViewBuilder
    private func productRow(_ p: Product) -> some View {
        HStack(spacing: 14) {
            VStack(alignment: .leading, spacing: 2) {
                Text(p.displayName.isEmpty ? p.id : p.displayName)
                    .font(.system(size: 16, weight: .bold, design: .rounded))
                Text(p.description.isEmpty ? "Image credits" : p.description)
                    .font(.system(size: 12)).foregroundStyle(.secondary)
            }
            Spacer()
            Button {
                Task { await buy(p) }
            } label: {
                if busy == p.id { ProgressView() }
                else {
                    Text(p.displayPrice)
                        .font(.system(size: 14, weight: .bold))
                        .padding(.horizontal, 16).padding(.vertical, 9)
                        .background(Color(hex: "#ff1493"))
                        .foregroundStyle(.white)
                        .clipShape(Capsule())
                }
            }
            .buttonStyle(.plain)
            .disabled(busy != nil)
        }
        .padding(14)
        .background(Color(.systemBackground))
        .clipShape(RoundedRectangle(cornerRadius: 14))
        .overlay(RoundedRectangle(cornerRadius: 14).stroke(Color(hex: "#f3c6dd"), lineWidth: 2))
    }

    // MARK: -- StoreKit

    private func load() async {
        loadFailed = false
        // Product.products(for:) does NOT throw for ids it can't resolve — it
        // silently drops them, and an all-unresolved list comes back as an
        // EMPTY array. Without this guard the view stays on the spinner
        // forever (empty products + no error). StoreKit is also flaky right
        // after launch, so retry briefly before declaring failure.
        for attempt in 0..<3 {
            do {
                let fetched = try await Product.products(for: Self.productIDs)
                if !fetched.isEmpty {
                    products = fetched.sorted { $0.price < $1.price }
                    break
                }
            } catch {}
            if attempt < 2 { try? await Task.sleep(for: .seconds(attempt + 1)) }
        }
        if products.isEmpty { loadFailed = true }
        // Silently re-post current entitlements so subscription RENEWALS land
        // server-side (idempotent per transaction id) — this is what keeps the
        // membership "active" month after month without the parent doing
        // anything.
        for await entitlement in Transaction.currentEntitlements {
            if case .verified(let tx) = entitlement, tx.productType == .autoRenewable {
                _ = await api.iapVerify(jws: entitlement.jwsRepresentation,
                                        productId: tx.productID,
                                        transactionId: String(tx.id))
            }
        }
        balance = await api.storeBalance()
        entitlement = await api.storeEntitlement()
    }

    private func buy(_ p: Product) async {
        busy = p.id
        defer { busy = nil }
        do {
            let result = try await p.purchase()
            switch result {
            case .success(let verification):
                guard case .verified(let tx) = verification else { return }
                let credited = await api.iapVerify(jws: verification.jwsRepresentation,
                                                   productId: tx.productID,
                                                   transactionId: String(tx.id))
                await tx.finish()
                balance = await api.storeBalance()
                entitlement = await api.storeEntitlement()
                if let credited, credited > 0 { note = "Added ⭐\(credited). Thank you!" }
            case .userCancelled, .pending: break
            @unknown default: break
            }
        } catch {
            note = "Purchase didn't complete: \(error.localizedDescription)"
        }
    }

    private func redeem() async {
        let code = couponCode.trimmingCharacters(in: .whitespaces)
        guard !code.isEmpty else { return }
        redeeming = true
        defer { redeeming = false }
        do {
            let r = try await api.storeRedeem(code: code)
            balance = r.balance
            couponCode = ""
            note = "Added ⭐\(r.credited). Enjoy!"
        } catch let APIError.badStatus(_, body) {
            note = body.contains("already used") ? "You've already used this code."
                 : body.contains("expired") ? "That code has expired."
                 : "That code isn't valid."
        } catch {
            note = "Couldn't redeem: \(error.localizedDescription)"
        }
    }

    /// Re-sends current entitlements/unfinished transactions to the server —
    /// idempotent there, so this is always safe.
    private func restore() async {
        try? await AppStore.sync()
        for await entitlement in Transaction.currentEntitlements {
            if case .verified(let tx) = entitlement {
                _ = await api.iapVerify(jws: entitlement.jwsRepresentation,
                                        productId: tx.productID,
                                        transactionId: String(tx.id))
            }
        }
        balance = await api.storeBalance()
        note = "Purchases restored."
    }
}

File: android-native/app/src/main/java/io/andrewpeterson/myworld/ui/parent/StoreView.kt
package io.andrewpeterson.myworld.ui.parent

import android.content.Intent
import android.net.Uri
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import io.andrewpeterson.myworld.LocalAppContainer
import io.andrewpeterson.myworld.net.ApiClient
import io.andrewpeterson.myworld.net.StoreCatalog
import io.andrewpeterson.myworld.net.storeCatalog
import io.andrewpeterson.myworld.net.storeRedeem
import kotlinx.coroutines.launch
import io.andrewpeterson.myworld.ui.theme.Brand
import io.andrewpeterson.myworld.ui.theme.hexColor

/**
 * Credits & membership — port of `Parent/StoreView.swift`: balance, tier,
 * voice budget, and NATIVE Google Play purchases for the three memberships
 * and credit packs (verify-before-acknowledge; server is source of truth).
 * On devices without Play services (Fire tablets) the buttons hand off to
 * the secure web store instead — that path works everywhere, permanently.
 */
@Composable
fun StoreView(onDismiss: () -> Unit) {
    val c = LocalAppContainer.current
    val context = LocalContext.current
    var catalog by remember { mutableStateOf<StoreCatalog?>(null) }
    var error by remember { mutableStateOf<String?>(null) }
    var showShop by remember { mutableStateOf(false) }
    var coupon by remember { mutableStateOf("") }
    var couponNote by remember { mutableStateOf<String?>(null) }
    val scope = androidx.compose.runtime.rememberCoroutineScope()

    val billingAvailable by c.billing.available.collectAsState()
    val billingNote by c.billing.note.collectAsState()
    val purchaseTick by c.billing.purchaseTick.collectAsState()
    val activity = context as? android.app.Activity

    LaunchedEffect(Unit) { c.billing.start() }
    LaunchedEffect(purchaseTick) {
        try { catalog = c.api.storeCatalog() }
        catch (e: Exception) { if (catalog == null) error = "Could not load the store: ${e.message}" }
        // Warm the Word Shop's catalog cache while the parent is still here,
        // so "Shop words for the board" opens instantly (iOS parity).
        if (purchaseTick == 0) {
            io.andrewpeterson.myworld.storage.ShopCatalogCache.refresh(context, c.auth.childSlug, c.api)
        }
    }

    fun openWebStore() {
        val url = "${ApiClient.ORIGIN}/store.html?child=${c.api.esc(c.auth.childSlug)}"
        try { context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url))) } catch (_: Exception) {}
    }

    Dialog(onDismissRequest = onDismiss, properties = DialogProperties(usePlatformDefaultWidth = false)) {
        Column(
            Modifier.fillMaxWidth(0.94f)
                .background(Color.White, RoundedCornerShape(24.dp))
                .padding(22.dp)
                .verticalScroll(rememberScrollState()),
        ) {
            Text("Credits & Store", fontSize = 22.sp, fontWeight = FontWeight.Bold, color = Brand.pinkDeep)
            Spacer(Modifier.height(12.dp))

            val cat = catalog
            when {
                error != null -> Text(error!!, fontSize = 13.sp, color = Color(0xFFDC2626))
                cat == null -> Text("Loading…", fontSize = 14.sp, color = Brand.muted)
                else -> {
                    // Balance + tier headline.
                    Column(
                        Modifier.fillMaxWidth()
                            .background(hexColor("#fdf2f8"), RoundedCornerShape(16.dp))
                            .padding(16.dp),
                    ) {
                        Row {
                            Column(Modifier.weight(1f)) {
                                Text("Image credits", fontSize = 12.sp, color = Brand.muted)
                                Text("${cat.balance ?: 0}", fontSize = 30.sp,
                                    fontWeight = FontWeight.Black, color = Brand.pinkDeep)
                            }
                            Column {
                                Text("Membership", fontSize = 12.sp, color = Brand.muted)
                                Text(cat.entitlement?.label ?: "Free", fontSize = 18.sp,
                                    fontWeight = FontWeight.Bold, color = Brand.ink)
                            }
                        }
                        cat.entitlement?.voice?.let { v ->
                            Spacer(Modifier.height(8.dp))
                            val capText = v.cap?.let { "${it / 1000}k" } ?: "∞"
                            Text("Voice this month: ${v.used / 1000}k of $capText characters",
                                fontSize = 12.sp, color = Brand.muted)
                        }
                    }
                    Spacer(Modifier.height(14.dp))

                    // Tier ladder — native Play purchase when available.
                    Text("MEMBERSHIPS", fontSize = 11.sp, fontWeight = FontWeight.Black, color = Brand.muted)
                    Spacer(Modifier.height(6.dp))
                    cat.subscriptions.forEach { s ->
                        val active = cat.entitlement?.tier?.let { s.sku.startsWith(it) } == true
                        Row(
                            Modifier.fillMaxWidth().padding(vertical = 6.dp),
                            verticalAlignment = androidx.compose.ui.Alignment.CenterVertically,
                        ) {
                            Column(Modifier.weight(1f)) {
                                Text(s.label.ifEmpty { s.sku }, fontSize = 15.sp,
                                    fontWeight = FontWeight.Bold,
                                    color = if (active) Brand.pinkDeep else Brand.ink)
                                Text("${s.creditsPerPeriod} image credits each month",
                                    fontSize = 12.sp, color = Brand.muted)
                            }
                            when {
                                active -> Text("Current", fontSize = 13.sp,
                                    fontWeight = FontWeight.SemiBold, color = hexColor("#10b981"))
                                billingAvailable && activity != null -> TextButton(onClick = {
                                    c.billing.launchPurchase(activity, s.sku)
                                }) {
                                    Text(c.billing.price(s.sku) ?: ("$" + "%.2f/mo".format(s.cents / 100.0)),
                                        fontSize = 13.sp, fontWeight = FontWeight.Bold, color = Brand.pinkDeep)
                                }
                                else -> Text("$" + "%.2f/mo".format(s.cents / 100.0),
                                    fontSize = 13.sp, fontWeight = FontWeight.SemiBold, color = Brand.muted)
                            }
                        }
                    }
                    if (cat.subscriptions.isEmpty()) {
                        Text("Memberships from $4.99/month unlock speech-to-text, auto-teach, reporting, and styled tiles.",
                            fontSize = 12.sp, color = Brand.muted)
                    }

                    // Credit packs — native buy buttons when Play is present.
                    if (billingAvailable && cat.packs.isNotEmpty()) {
                        Spacer(Modifier.height(12.dp))
                        Text("CREDIT PACKS", fontSize = 11.sp, fontWeight = FontWeight.Black, color = Brand.muted)
                        Spacer(Modifier.height(4.dp))
                        cat.packs.forEach { p ->
                            Row(
                                Modifier.fillMaxWidth().padding(vertical = 2.dp),
                                verticalAlignment = androidx.compose.ui.Alignment.CenterVertically,
                            ) {
                                Text("⭐ ${p.credits} — ${p.label.ifEmpty { p.sku }}",
                                    fontSize = 14.sp, fontWeight = FontWeight.SemiBold,
                                    color = Brand.ink, modifier = Modifier.weight(1f))
                                TextButton(onClick = {
                                    if (activity != null) c.billing.launchPurchase(activity, p.sku)
                                }) {
                                    Text(c.billing.price(p.sku) ?: ("$" + "%.2f".format(p.cents / 100.0)),
                                        fontSize = 13.sp, fontWeight = FontWeight.Bold, color = Brand.pinkDeep)
                                }
                            }
                        }
                    }

                    billingNote?.let {
                        Spacer(Modifier.height(8.dp))
                        Text(it, fontSize = 12.sp, fontWeight = FontWeight.SemiBold, color = hexColor("#047857"))
                    }
                }
            }

            Spacer(Modifier.height(16.dp))
            Button(
                onClick = { showShop = true },
                colors = ButtonDefaults.buttonColors(containerColor = hexColor("#ad1457")),
                modifier = Modifier.fillMaxWidth().height(48.dp),
            ) {
                Text("🛍 Shop words for the board", fontWeight = FontWeight.Bold)
            }
            Spacer(Modifier.height(10.dp))
            Button(
                onClick = { openWebStore() },
                colors = ButtonDefaults.buttonColors(containerColor = Brand.pink),
                modifier = Modifier.fillMaxWidth().height(48.dp),
            ) {
                Text(if (billingAvailable) "Manage billing on the web" else "Buy on the secure web store",
                    fontWeight = FontWeight.Bold)
            }

            Spacer(Modifier.height(10.dp))
            Row(verticalAlignment = androidx.compose.ui.Alignment.CenterVertically) {
                androidx.compose.material3.OutlinedTextField(
                    value = coupon, onValueChange = { coupon = it },
                    label = { Text("Have a code?") }, singleLine = true,
                    modifier = Modifier.weight(1f),
                )
                TextButton(onClick = {
                    val code = coupon.trim()
                    if (code.isEmpty()) return@TextButton
                    scope.launch {
                        couponNote = try {
                            val r = c.api.storeRedeem(code)
                            coupon = ""
                            try { catalog = c.api.storeCatalog() } catch (_: Exception) {}
                            "⭐ ${r.credited} credits added!"
                        } catch (e: Exception) { "Couldn't redeem: ${e.message}" }
                    }
                }) { Text("Redeem", fontWeight = FontWeight.Bold, color = Brand.pinkDeep) }
            }
            couponNote?.let { Text(it, fontSize = 12.sp, color = Brand.muted) }
            if (billingAvailable) {
                TextButton(onClick = { c.billing.restorePurchases() }, modifier = Modifier.fillMaxWidth()) {
                    Text("Restore purchases", fontSize = 13.sp, color = Brand.pinkDeep)
                }
            }
            Text(
                if (billingAvailable)
                    "Prices are billed through Google Play. Purchases made on the web appear here right away too — one wallet everywhere."
                else
                    "This device doesn't have Google Play (that's normal on Fire tablets), so purchases open the secure web store instead. Credits bought there land on the board right away.",
                fontSize = 11.sp, color = Brand.muted,
                modifier = Modifier.padding(top = 8.dp),
            )
            Row {
                TextButton(onClick = {
                    try { context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("${ApiClient.ORIGIN}/terms"))) } catch (_: Exception) {}
                }) { Text("Terms of Service", fontSize = 11.sp, color = Brand.muted) }
                TextButton(onClick = {
                    try { context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("${ApiClient.ORIGIN}/privacy"))) } catch (_: Exception) {}
                }) { Text("Privacy Policy", fontSize = 11.sp, color = Brand.muted) }
            }
            TextButton(onClick = onDismiss, modifier = Modifier.fillMaxWidth()) {
                Text("Close", color = Brand.muted)
            }
        }
    }

    if (showShop) WordShopView { showShop = false }
}
